examples/algebra/multipoly/multipolyScript.sml

```diff
@@ -0,0 +1,2232 @@
+
+Theorem GBAG_IMAGE_PARTITION:
+  AbelianMonoid g /\ FINITE s ==>
+  !b. FINITE_BAG b ==>
+    IMAGE f (SET_OF_BAG b) SUBSET G /\
+    (!x. BAG_IN x b ==> ?P. P IN s /\ P x) /\
+    (!x P1 P2. BAG_IN x b /\ P1 IN s /\ P2 IN s /\ P1 x /\ P2 x ==> P1 = P2)
+  ==>
+    GBAG g (BAG_IMAGE (λP. GBAG g (BAG_IMAGE f (BAG_FILTER P b))) (BAG_OF_SET s)) =
+    GBAG g (BAG_IMAGE f b)
+Proof
+  strip_tac
+  \\ ho_match_mp_tac STRONG_FINITE_BAG_INDUCT
+  \\ simp[]
+  \\ conj_tac
+  >- (
+    irule IMP_GBAG_EQ_ID
+    \\ simp[BAG_EVERY]
+    \\ rw[]
+    \\ imp_res_tac BAG_IN_BAG_IMAGE_IMP
+    \\ fs[] )
+  \\ rpt strip_tac
+  \\ fs[SET_OF_BAG_INSERT]
+  \\ `∃P. P IN s /\ P e` by metis_tac[]
+  \\ `∃s0. s = P INSERT s0 /\ P NOTIN s0` by metis_tac[DECOMPOSITION]
+  \\ BasicProvers.VAR_EQ_TAC
+  \\ simp[BAG_OF_SET_INSERT_NON_ELEMENT]
+  \\ DEP_REWRITE_TAC[BAG_IMAGE_FINITE_INSERT]
+  \\ qpat_x_assum`_ ⇒ _`mp_tac
+  \\ impl_tac >- metis_tac[]
+  \\ strip_tac
+  \\ conj_tac >- metis_tac[FINITE_INSERT, FINITE_BAG_OF_SET]
+  \\ qmatch_goalsub_abbrev_tac`BAG_IMAGE ff (BAG_OF_SET s0)`
+  \\ `BAG_IMAGE ff (BAG_OF_SET s0) =
+      BAG_IMAGE (\P. GBAG g (BAG_IMAGE f (BAG_FILTER P b))) (BAG_OF_SET s0)`
+  by (
+    irule BAG_IMAGE_CONG
+    \\ simp[Abbr`ff`]
+    \\ rw[]
+    \\ metis_tac[IN_INSERT] )
+  \\ simp[Abbr`ff`]
+  \\ pop_assum kall_tac
+  \\ rpt(first_x_assum(qspec_then`ARB`kall_tac))
+  \\ pop_assum mp_tac
+  \\ simp[BAG_OF_SET_INSERT_NON_ELEMENT]
+  \\ DEP_REWRITE_TAC[GITBAG_INSERT_THM]
+  \\ fs[AbelianMonoid_def]
+  \\ conj_asm1_tac >- fs[SUBSET_DEF, PULL_EXISTS]
+  \\ conj_asm1_tac >- (
+    fs[SUBSET_DEF, PULL_EXISTS]
+    \\ rw[] \\ irule GITBAG_in_carrier
+    \\ fs[SUBSET_DEF, PULL_EXISTS, AbelianMonoid_def] )
+  \\ simp[]
+  \\ DEP_REWRITE_TAC[GITBAG_INSERT_THM]
+  \\ simp[]
+  \\ conj_asm1_tac
+  >- (
+    simp[AbelianMonoid_def]
+    \\ irule GITBAG_in_carrier
+    \\ simp[AbelianMonoid_def] )
+  \\ simp[]
+  \\ DEP_ONCE_REWRITE_TAC[GITBAG_GBAG] \\ simp[] \\ strip_tac
+  \\ DEP_ONCE_REWRITE_TAC[GITBAG_GBAG] \\ simp[]
+  \\ DEP_ONCE_REWRITE_TAC[GITBAG_GBAG] \\ simp[]
+  \\ DEP_REWRITE_TAC[monoid_assoc]
+  \\ simp[]
+  \\ conj_tac >- ( irule GBAG_in_carrier \\ simp[] )
+  \\ irule EQ_SYM
+  \\ irule GITBAG_GBAG
+  \\ simp[]
+QED
+
+Theorem GBAG_PARTITION:
+  AbelianMonoid g /\ FINITE s /\ FINITE_BAG b /\ SET_OF_BAG b SUBSET G /\
+    (!x. BAG_IN x b ==> ?P. P IN s /\ P x) /\
+    (!x P1 P2. BAG_IN x b /\ P1 IN s /\ P2 IN s /\ P1 x /\ P2 x ==> P1 = P2)
+  ==>
+    GBAG g (BAG_IMAGE (λP. GBAG g (BAG_FILTER P b)) (BAG_OF_SET s)) = GBAG g b
+Proof
+  strip_tac
+  \\ `!P. FINITE_BAG (BAG_FILTER P b)` by metis_tac[FINITE_BAG_FILTER]
+  \\ `GBAG g b = GBAG g (BAG_IMAGE I b)` by metis_tac[BAG_IMAGE_FINITE_I]
+  \\ pop_assum SUBST1_TAC
+  \\ `(λP. GBAG g (BAG_FILTER P b)) = λP. GBAG g (BAG_IMAGE I (BAG_FILTER P b))`
+  by simp[FUN_EQ_THM]
+  \\ pop_assum SUBST1_TAC
+  \\ irule GBAG_IMAGE_PARTITION
+  \\ simp[]
+  \\ metis_tac[]
+QED
+
+Theorem ring_mult_lsum:
```

Comment:
Indentation here is gross.

Reply:
I don't see the issue - what's the proposed change?

Reply:
oh maybe GitHub was showing me the wrong quote - do you mean the statement of ring_mult_lsum? I will improve that.